src/hybrid/parse_hybrid.c

```diff
@@ -70,41 +72,282 @@ static int parseSearchSubquery(ArgsCursor *ac, AREQ *searchRequest, QueryError *
+    } else {
+      QueryError_SetWithUserDataFmt(status, QUERY_EPARSEARGS, "Unknown parameter", " `%s` in RANGE", current);
+      return REDISMODULE_ERR;
+    }
+  }
+  if (!hasRadius) {
+    QueryError_SetError(status, QUERY_EPARSEARGS, "Missing RADIUS parameter");
+    return REDISMODULE_ERR;
+  }
+  return REDISMODULE_OK;
+}
+static int parseFilterClause(ArgsCursor *ac, AREQ *vectorRequest, QueryError *status) {
+  // FILTER is in our scope, advance and process it
+  AC_Advance(ac);
+  vectorRequest->query = AC_GetStringNC(ac, NULL);
+  return REDISMODULE_OK;
+}
+
 static int parseVectorSubquery(ArgsCursor *ac, AREQ *vectorRequest, QueryError *status) {
-  // Check if VSIM parameter is present
-  if (!AC_AdvanceIfMatch(ac, "VSIM")) {
+  const char *current;
+  if (AC_GetString(ac, &current, NULL, AC_F_NOADVANCE) != AC_OK || strcasecmp("VSIM", current)) {
     QueryError_SetError(status, QUERY_ESYNTAX, "VSIM parameter is required");
     return REDISMODULE_ERR;
   }
+  AC_Advance(ac);
 
-  // Advance the cursor until we encounter one of the specified keywords or reach the end
-  while (!AC_IsAtEnd(ac)) {
-    const char *cur;
-    if (AC_GetString(ac, &cur, NULL, AC_F_NOADVANCE) != AC_OK) {
-      break;
-    }
+  // Allocate ParsedVectorQuery
+  ParsedVectorQuery *pvq = rm_calloc(1, sizeof(ParsedVectorQuery));
+
+  // Parse vector field and blob
+  if (AC_GetString(ac, &pvq->fieldName, NULL, 0) != AC_OK) {
+    QueryError_SetError(status, QUERY_ESYNTAX, "Missing vector field name");
+    goto error;
+  }
+
+  const char *vectorParam;
+  if (AC_GetString(ac, &vectorParam, NULL, 0) != AC_OK ) {
+    QueryError_SetError(status, QUERY_ESYNTAX, "Missing vector blob");
+    goto error;
+  }
+
+  if (vectorParam[0] == '$') {
+    // PARAMETER CASE: store parameter name for later resolution
+    vectorParam++;  // Skip '$'
+    pvq->vector = vectorParam;  // Parameter name
+    pvq->vectorLen = strlen(vectorParam);
+    pvq->isParameter = true;
+  } else {
+    // DIRECT VECTOR CASE: store the actual vector data
+    pvq->vector = vectorParam;  // Actual vector data
+    pvq->vectorLen = strlen(vectorParam);
+    pvq->isParameter = false;
+  }
+  if (AC_IsAtEnd(ac)) goto final;
+
+  // Initialize QueryAttribute array for attributes like YIELD_DISTANCE_AS
+  pvq->attributes = array_new(QueryAttribute, 0);
 
-    // Check if current argument is one of the keywords that should stop parsing
-    if (!strcasecmp(cur, "COMBINE") ||
-        !strcasecmp(cur, "LOAD") ||
-        !strcasecmp(cur, "GROUPBY") ||
-        !strcasecmp(cur, "APPLY") ||
-        !strcasecmp(cur, "SORTBY") ||
-        !strcasecmp(cur, "FILTER") ||
-        !strcasecmp(cur, "LIMIT") ||
-        !strcasecmp(cur, "PARAMS") ||
-        !strcasecmp(cur, "EXPLAINSCORE") ||
-        !strcasecmp(cur, "TIMEOUT")) {
-      // Found a keyword that should stop parsing, don't advance past it
-      break;
+  if (AC_GetString(ac, &current, NULL, AC_F_NOADVANCE) != AC_OK && !AC_IsAtEnd(ac)) {
```

Comment:
Consider using `AdvanceIfMatch` for KNN, then RANGE, then FILTER. I think it will be a bit cleaner